nextjs/src/app/(logged-in)/data-sources/create/connect/[externalDataSourceType]/page.tsx

```diff
@@ -1210,22 +1231,51 @@ export default function Page({
               name="editablegooglesheets.spreadsheetId"
               render={({ field }) => (
                 <FormItem>
-                  <FormLabel>Spreadsheet ID</FormLabel>
+                  <FormLabel>Google Sheets URL</FormLabel>
                   <FormControl>
-                    {/* @ts-ignore */}
                     <Input
-                      placeholder="1MEDFli9uakvmf_wGghJZtZg2AvF2xybGtiaG7OX1mmg"
-                      {...field}
+                      placeholder="https://docs.google.com/spreadsheets/d/1MEDFli9uakvmf_wGghJZtZg2AvF2xybGtiaG7OX1mmg/edit#gid=0"
+                      value={sheetUrl}
+                      onChange={async (e) => {
+                        const url = e.target.value
+                        setSheetUrl(url)
+                        try {
+                          const match = url.match(/\/d\/([a-zA-Z0-9-_]+)/)
+                          if (match && match[1]) {
+                            const spreadsheetId = match[1]
+                            form.setValue(
+                              'editablegooglesheets.spreadsheetId',
+                              spreadsheetId
+                            )
+                            setLoadingSheets(true)
+                            const oauthCredentials = form.getValues(
+                              'editablegooglesheets.oauthCredentials'
+                            )
+                            if (oauthCredentials) {
+                              const sheets =
+                                await fetchSheetNamesUsingCredentials(
+                                  spreadsheetId,
+                                  oauthCredentials
+                                )
+                              setSheetNames(sheets)
+                            } else {
+                              throw new Error('OAuth credentials not available')
+                            }
+
+                            setLoadingSheets(false)
+                          }
+                        } catch (err) {
+                          setFetchError(
+                            'Failed to fetch sheet names. Please check the URL or credentials.'
+                          )
+                          setLoadingSheets(false)
+                        }
+                      }}
```

Comment:
Looks good but can we define this function elsewhere so the UI bit of the file is kept tidy JSX? Inspired by @ev-sc's recent changes to the map code.

Reply:
good shout, done